File: sdk-ts/src/llm-loop.ts
/**
 * Built-in LLM loop for pipeline mode when no onMessage handler is provided.
 *
 * Uses a pluggable ``LLMProvider`` interface so callers can supply OpenAI,
 * Anthropic, Gemini, or any custom provider.  The default provider is
 * ``OpenAILLMProvider`` which preserves full backward compatibility.
 */

import type { ToolDefinition } from './types';
import { getLogger } from './logger';
import { validateWebhookUrl } from './server';
import { SPAN_TOOL, withSpan } from './observability/tracing';

// ---------------------------------------------------------------------------
// Tool execution — pluggable policy
// ---------------------------------------------------------------------------

/**
 * Minimal interface for recording LLM usage chunks.
 * Avoids a circular import from metrics.ts.
 */
export interface LlmUsageRecorder {
  recordLlmUsage(
    provider: string,
    model: string,
    inputTokens: number,
    outputTokens: number,
    cacheReadTokens?: number,
    cacheCreationTokens?: number,
  ): void;
}

const DEFAULT_TOOL_MAX_RETRIES = 2;
const DEFAULT_TOOL_RETRY_DELAY_MS = 500;
const DEFAULT_TOOL_TIMEOUT_MS = 10_000;
const TOOL_MAX_RESPONSE_BYTES = 1 * 1024 * 1024;

/**
 * Pluggable tool executor — mirrors the Python ``ToolExecutor`` in
 * ``sdk-py/getpatter/services/tool_executor.py``.
 *
 * Implementors receive a fully-resolved ``ToolDefinition`` (handler +/ webhook
 * URL already validated by the SDK) and MUST return a JSON-stringifiable
 * result. Errors should be returned as JSON like
 * ``{ error: "...", fallback: true }`` rather than thrown.
 */
export interface ToolExecutor {
  execute(
    toolDef: ToolDefinition,
    args: Record<string, unknown>,
    callContext: Record<string, unknown>,
  ): Promise<string>;
}

export interface DefaultToolExecutorOptions {
  /** Total attempts = maxRetries + 1. Default: 2 (i.e. 3 attempts). */
  maxRetries?: number;
  /** Delay between attempts, in ms. */
  retryDelayMs?: number;
  /** Per-request timeout for webhook calls, in ms. */
  requestTimeoutMs?: number;
}

/**
 * Default executor — webhook with retry/fallback and local handler preference.
 *
 * This is the out-of-the-box behavior and is 1:1 equivalent to the previous
 * inline logic in ``LLMLoop.executeTool``.
 */
export class DefaultToolExecutor implements ToolExecutor {
  private readonly maxRetries: number;
  private readonly retryDelayMs: number;
  private readonly requestTimeoutMs: number;

  constructor(opts: DefaultToolExecutorOptions = {}) {
    this.maxRetries = opts.maxRetries ?? DEFAULT_TOOL_MAX_RETRIES;
    this.retryDelayMs = opts.retryDelayMs ?? DEFAULT_TOOL_RETRY_DELAY_MS;
    this.requestTimeoutMs = opts.requestTimeoutMs ?? DEFAULT_TOOL_TIMEOUT_MS;
  }

  async execute(
    toolDef: ToolDefinition,
    args: Record<string, unknown>,
    callContext: Record<string, unknown>,
  ): Promise<string> {
    // Prefer local handler.
    if (toolDef.handler) {
      try {
        return await toolDef.handler(args, callContext);
      } catch (e) {
        return JSON.stringify({
          error: `Tool handler error: ${String(e)}`,
          fallback: true,
        });
      }
    }

    // Fall back to webhook with retry/backoff.
    if (toolDef.webhookUrl) {
      try {
        validateWebhookUrl(toolDef.webhookUrl);
      } catch (e) {
        return JSON.stringify({ error: `Tool webhook URL rejected: ${String(e)}` });
      }
      const callId = typeof callContext.call_id === 'string' ? callContext.call_id : '';
      return await withSpan(
        SPAN_TOOL,
        {
          'patter.tool.name': toolDef.name,
          'patter.tool.transport': 'webhook',
          'patter.call.id': callId,
        },
        async (span) => {
          const totalAttempts = this.maxRetries + 1;
          for (let attempt = 0; attempt < totalAttempts; attempt++) {
            span.setAttribute('patter.tool.attempt', attempt + 1);
            try {
              const resp = await fetch(toolDef.webhookUrl!, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                  tool: toolDef.name,
                  arguments: args,
                  ...callContext,
                  attempt: attempt + 1,
                }),
                signal: AbortSignal.timeout(this.requestTimeoutMs),
              });
              if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
              const result = JSON.stringify(await resp.json());
              if (result.length > TOOL_MAX_RESPONSE_BYTES) {
                return JSON.stringify({
                  error: `Webhook response too large: ${result.length} bytes (max ${TOOL_MAX_RESPONSE_BYTES})`,
                  fallback: true,
                });
              }
              return result;
            } catch (e) {
              if (attempt < totalAttempts - 1) {
                getLogger().warn(
                  `Tool webhook '${toolDef.name}' failed (attempt ${attempt + 1}), retrying: ${String(e)}`,
                );
                await new Promise<void>((r) => setTimeout(r, this.retryDelayMs));
              } else {
                span.recordException(e);
                return JSON.stringify({
                  error: `Tool failed after ${totalAttempts} attempts: ${String(e)}`,
                  fallback: true,
                });
              }
            }
          }
          // Unreachable — the for-loop always returns.
          return JSON.stringify({
            error: `Tool '${toolDef.name}' exited retry loop unexpectedly`,
            fallback: true,
          });
        },
      );
    }

    return JSON.stringify({
      error: `No handler or webhookUrl for tool '${toolDef.name}'`,
      fallback: true,
    });
  }
}

// ---------------------------------------------------------------------------
// Provider interface
// ---------------------------------------------------------------------------

/** A single streaming chunk yielded by an LLM provider. */
export interface LLMChunk {
  type: 'text' | 'tool_call' | 'done' | 'usage';
  content?: string;
  index?: number;
  id?: string;
  name?: string;
  arguments?: string;
  // Fix 10: usage chunk fields (emitted by providers that expose token counts)
  inputTokens?: number;
  outputTokens?: number;
  cacheReadInputTokens?: number;
  cacheCreationInputTokens?: number;
}

/**
 * Interface that any LLM provider must satisfy.
 *
 * Implementors yield streaming ``LLMChunk`` objects:
 * - ``{ type: "text", content: "..." }`` — a text token.
 * - ``{ type: "tool_call", index, id?, name?, arguments? }`` — a (partial) tool
 *   invocation.  Chunks with the same ``index`` are concatenated.
 * - ``{ type: "done" }`` — signals the end of the stream (optional).
 */
export interface LLMProvider {
  stream(
    messages: Array<Record<string, unknown>>,
    tools?: Array<Record<string, unknown>> | null,
  ): AsyncGenerator<LLMChunk, void, unknown>;
}

// ---------------------------------------------------------------------------
// Built-in OpenAI provider
// ---------------------------------------------------------------------------

/** LLM provider backed by OpenAI Chat Completions (streaming). */
export class OpenAILLMProvider implements LLMProvider {
  private readonly apiKey: string;
  readonly model: string;

  constructor(apiKey: string, model: string) {
    this.apiKey = apiKey;
    this.model = model;
  }

  async *stream(
    messages: Array<Record<string, unknown>>,
    tools?: Array<Record<string, unknown>> | null,
  ): AsyncGenerator<LLMChunk, void, unknown> {
    const body: Record<string, unknown> = {
      model: this.model,
      messages,
      stream: true,
      // Ask OpenAI to include a final usage chunk so we can attribute token
      // cost. Without this the dashboard shows LLM cost = 0 for OpenAI.
      stream_options: { include_usage: true },
    };
    if (tools) {
      body.tools = tools;
    }

    const response = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(30_000),
    });

    if (!response.ok) {
      const errText = await response.text();
      getLogger().error(`LLM API error: ${response.status} ${errText}`);
      return;
    }

    const reader = response.body?.getReader();
    if (!reader) return;

    const decoder = new TextDecoder();
    let buffer = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed || !trimmed.startsWith('data: ')) continue;
        const data = trimmed.slice(6);
        if (data === '[DONE]') continue;

        let chunk: {
          choices?: Array<{
            delta?: {
              content?: string;
              tool_calls?: Array<{
                index: number;
                id?: string;
                function?: { name?: string; arguments?: string };
              }>;
            };
          }>;
          usage?: {
            prompt_tokens?: number;
            completion_tokens?: number;
            prompt_tokens_details?: { cached_tokens?: number };
          };
        };
        try {
          chunk = JSON.parse(data);
        } catch {
          continue;
        }

        // Final usage chunk arrives with choices=[] when stream_options
        // include_usage is set. Forward it for cost attribution.
        if (chunk.usage) {
          const cached = chunk.usage.prompt_tokens_details?.cached_tokens ?? 0;
          // OpenAI's prompt_tokens is the TOTAL input including cached tokens.
          // Subtract cached so inputTokens represents only the uncached portion
          // and calculateLlmCost doesn't bill cached tokens at the full rate.
          const uncachedInput = Math.max(0, (chunk.usage.prompt_tokens ?? 0) - cached);
          yield {
            type: 'usage',
            inputTokens: uncachedInput,
            outputTokens: chunk.usage.completion_tokens,
            cacheReadInputTokens: cached,
          };
        }

        const delta = chunk.choices?.[0]?.delta;
        if (!delta) continue;

        if (delta.content) {
          yield { type: 'text', content: delta.content };
        }

        if (delta.tool_calls) {
          for (const tc of delta.tool_calls) {
            yield {
              type: 'tool_call',
              index: tc.index,
              id: tc.id,
              name: tc.function?.name,
              arguments: tc.function?.arguments,
            };
          }
        }
      }
    }
  }
}

// ---------------------------------------------------------------------------
// Internal types
// ---------------------------------------------------------------------------

interface OpenAIMessage {
  role: string;
  content?: string | null;
  tool_calls?: Array<{
    id: string;
    type: string;
    function: { name: string; arguments: string };
  }>;
  tool_call_id?: string;
  [key: string]: unknown;
}

interface ToolCallAccumulator {
  id: string;
  name: string;
  arguments: string;
}

// ---------------------------------------------------------------------------
// LLM loop
// ---------------------------------------------------------------------------

export class LLMLoop {
  private readonly provider: LLMProvider;
  private readonly systemPrompt: string;
  private readonly tools: ToolDefinition[] | null;
  private readonly openaiTools: Array<{
    type: string;
    function: { name: string; description: string; parameters: Record<string, unknown> };
  }> | null;
  private readonly toolMap: Map<string, ToolDefinition>;
  private toolExecutor: ToolExecutor;
  // Fix 10: track provider/model so usage chunks can be attributed for billing.
  private readonly _providerName: string;
  private readonly _modelName: string;

  constructor(
    apiKey: string,
    model: string,
    systemPrompt: string,
    tools?: ToolDefinition[] | null,
    llmProvider?: LLMProvider,
  ) {
    this.provider = llmProvider ?? new OpenAILLMProvider(apiKey, model);
    this.systemPrompt = systemPrompt;
    // Derive a billing-friendly provider name. Prefer the static
    // ``providerKey`` (stable, matches pricing keys); fall back to the
    // class-name stripping heuristic for custom providers without it.
    if (llmProvider) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const key = (llmProvider.constructor as any)?.providerKey;
      if (key) {
        this._providerName = key;
      } else {
        const stripped = (llmProvider.constructor?.name ?? 'custom')
          .replace(/LLMProvider$/i, '')
          .replace(/LLM$/i, '')
          .replace(/Provider$/i, '')
          .toLowerCase();
        this._providerName = stripped || 'custom';
      }
    } else {
      this._providerName = 'openai';
    }
    this._modelName = model;
    this.tools = tools ?? null;
    this.toolExecutor = new DefaultToolExecutor();

    this.toolMap = new Map();
    this.openaiTools = null;

    if (this.tools && this.tools.length > 0) {
      this.openaiTools = [];
      for (const t of this.tools) {
        this.openaiTools.push({
          type: 'function',
          function: {
            name: t.name,
            description: t.description || '',
            parameters: t.parameters || { type: 'object', properties: {} },
          },
        });
        this.toolMap.set(t.name, t);
      }
    }
  }

  /**
   * Swap in a custom tool executor (e.g. different retry policy, metrics
   * wrapping, tenant-aware fan-out). The default is ``DefaultToolExecutor``.
   */
  setToolExecutor(executor: ToolExecutor): void {
    this.toolExecutor = executor;
  }

  /**
   * Stream LLM response tokens, handling tool calls automatically.
   * Yields text tokens as they arrive from the LLM.
   *
   * @param metrics Optional usage recorder — when provided, usage chunks
   *   from the provider are forwarded to {@link LlmUsageRecorder.recordLlmUsage}
   *   so token costs are included in the call cost breakdown (fix 10).
   */
  async *run(
    userText: string,
    history: Array<{ role: string; text: string }>,
    callContext: Record<string, unknown>,
    metrics?: LlmUsageRecorder,
  ): AsyncGenerator<string, void, unknown> {
    const messages = this.buildMessages(history, userText);
    const maxIterations = 10;

    for (let iter = 0; iter < maxIterations; iter++) {
      const toolCallsAccumulated = new Map<number, ToolCallAccumulator>();
      const textParts: string[] = [];
      let hasToolCalls = false;

      for await (const chunk of this.provider.stream(messages, this.openaiTools)) {
        if (chunk.type === 'text' && chunk.content) {
          textParts.push(chunk.content);
          yield chunk.content;
        } else if (chunk.type === 'usage') {
          // Fix 10: forward token usage to the metrics accumulator for billing.
          metrics?.recordLlmUsage(
            this._providerName,
            this._modelName,
            chunk.inputTokens ?? 0,
            chunk.outputTokens ?? 0,
            chunk.cacheReadInputTokens ?? 0,
            chunk.cacheCreationInputTokens ?? 0,
          );
        } else if (chunk.type === 'tool_call') {
          hasToolCalls = true;
          const idx = chunk.index ?? 0;
          if (!toolCallsAccumulated.has(idx)) {
            toolCallsAccumulated.set(idx, { id: '', name: '', arguments: '' });
          }
          const acc = toolCallsAccumulated.get(idx)!;
          if (chunk.id) acc.id = chunk.id;
          if (chunk.name) acc.name = chunk.name;
          if (chunk.arguments) acc.arguments += chunk.arguments;
        }
      }

      if (!hasToolCalls) return;

      // Execute tool calls and add results to messages
      const assistantMsg: OpenAIMessage = {
        role: 'assistant',
        content: textParts.join('') || null,
        tool_calls: [],
      };

      const sortedIndices = [...toolCallsAccumulated.keys()].sort((a, b) => a - b);
      for (const idx of sortedIndices) {
        const tc = toolCallsAccumulated.get(idx)!;
        assistantMsg.tool_calls!.push({
          id: tc.id,
          type: 'function',
          function: { name: tc.name, arguments: tc.arguments },
        });
      }
      messages.push(assistantMsg);

      for (const tcData of assistantMsg.tool_calls!) {
        const toolName = tcData.function.name;
        let args: Record<string, unknown>;
        try {
          args = JSON.parse(tcData.function.arguments);
        } catch {
          args = {};
        }

        const result = await this.executeTool(toolName, args, callContext);
        messages.push({
          role: 'tool',
          tool_call_id: tcData.id,
          content: result,
        });
      }
    }

    getLogger().warn(`LLM loop hit max iterations (${maxIterations})`);
  }

  private async executeTool(
    toolName: string,
    args: Record<string, unknown>,
    callContext: Record<string, unknown>,
  ): Promise<string> {
    const toolDef = this.toolMap.get(toolName);
    if (!toolDef) {
      return JSON.stringify({ error: `Unknown tool: ${toolName}` });
    }
    return this.toolExecutor.execute(toolDef, args, callContext);
  }

  private buildMessages(
    history: Array<{ role: string; text: string }>,
    userText: string,
  ): OpenAIMessage[] {
    const messages: OpenAIMessage[] = [
      { role: 'system', content: this.systemPrompt },
    ];

    for (const entry of history) {
      messages.push({
        role: entry.role === 'assistant' ? 'assistant' : 'user',
        content: entry.text,
      });
    }

    messages.push({ role: 'user', content: userText });
    return messages;
  }
}

File: sdk-py/getpatter/services/llm_loop.py
"""Built-in LLM loop for pipeline mode when no on_message handler is provided.

Uses a pluggable ``LLMProvider`` protocol so callers can supply OpenAI,
Anthropic, Gemini, or any custom provider.  The default provider is
``OpenAILLMProvider`` which preserves full backward compatibility.
"""

from __future__ import annotations

__all__ = ["LLMLoop", "LLMProvider", "OpenAILLMProvider"]

import json
import logging
from typing import AsyncGenerator, AsyncIterator, Protocol, runtime_checkable

from getpatter.observability.tracing import SPAN_LLM, start_span

logger = logging.getLogger("getpatter")


# ---------------------------------------------------------------------------
# Provider protocol
# ---------------------------------------------------------------------------


@runtime_checkable
class LLMProvider(Protocol):
    """Protocol that any LLM provider must satisfy.

    Implementors yield streaming chunks as dicts.  Each chunk must include a
    ``"type"`` key:

    * ``{"type": "text", "content": "..."}`` — a text token.
    * ``{"type": "tool_call", "index": int, "id": str | None,
       "name": str | None, "arguments": str | None}`` — a (partial) tool
       invocation.  Chunks with the same ``index`` are concatenated.
    * ``{"type": "done"}`` — signals the end of the stream (optional).
    """

    async def stream(
        self,
        messages: list[dict],
        tools: list[dict] | None = None,
    ) -> AsyncIterator[dict]:
        """Yield streaming chunks for the given messages and tools."""
        ...  # pragma: no cover


# ---------------------------------------------------------------------------
# Built-in OpenAI provider
# ---------------------------------------------------------------------------


class OpenAILLMProvider:
    """LLM provider backed by OpenAI Chat Completions (streaming).

    Args:
        api_key: OpenAI API key.
        model: Chat model ID (e.g. ``"gpt-4o-mini"``).
    """

    def __init__(self, api_key: str, model: str) -> None:
        try:
            from openai import AsyncOpenAI
        except ImportError:
            raise ImportError(
                "The 'openai' package is required for the built-in OpenAI LLM "
                "provider. Install it with: pip install openai"
            )

        self._client = AsyncOpenAI(api_key=api_key)
        self._model = model

    async def stream(
        self,
        messages: list[dict],
        tools: list[dict] | None = None,
    ) -> AsyncIterator[dict]:
        """Yield normalised chunks from OpenAI Chat Completions.

        Emits a final ``{"type": "usage", ...}`` chunk when the upstream
        response includes a ``usage`` field (enabled via
        ``stream_options={"include_usage": True}``). Downstream callers
        use this to attribute real input/output token counts to the call
        instead of estimating from text length.
        """
        response = await self._client.chat.completions.create(
            model=self._model,
            messages=messages,
            tools=tools if tools else None,
            stream=True,
            stream_options={"include_usage": True},
        )

        last_usage = None
        async for chunk in response:
            # Usage chunks have empty ``choices`` and a populated ``usage``.
            usage = getattr(chunk, "usage", None)
            if usage is not None:
                last_usage = usage

            delta = chunk.choices[0].delta if chunk.choices else None
            if delta is None:
                continue

            if delta.content:
                yield {"type": "text", "content": delta.content}

            if delta.tool_calls:
                for tc in delta.tool_calls:
                    yield {
                        "type": "tool_call",
                        "index": tc.index,
                        "id": tc.id,
                        "name": tc.function.name if tc.function else None,
                        "arguments": tc.function.arguments if tc.function else None,
                    }

        if last_usage is not None:
            cache_read = 0
            details = getattr(last_usage, "prompt_tokens_details", None)
            if details is not None:
                cache_read = getattr(details, "cached_tokens", 0) or 0
            # OpenAI's prompt_tokens is the TOTAL input (uncached + cached).
            # Subtract cached so input_tokens represents only the uncached
            # portion and calculate_llm_cost doesn't bill cached tokens at
            # the full input rate (mirrors sdk-ts/llm-loop.ts:296-305).
            prompt_tokens = getattr(last_usage, "prompt_tokens", 0) or 0
            uncached_input = max(0, prompt_tokens - cache_read)
            yield {
                "type": "usage",
                "input_tokens": uncached_input,
                "output_tokens": getattr(last_usage, "completion_tokens", 0) or 0,
                "cache_read_tokens": cache_read,
            }


# ---------------------------------------------------------------------------
# LLM loop
# ---------------------------------------------------------------------------


class LLMLoop:
    """Streaming LLM with tool calling for pipeline mode.

    When ``agent.provider == "pipeline"`` and no ``on_message`` callback is
    provided, this class handles the LLM interaction internally.

    Args:
        openai_key: OpenAI API key (used when *llm_provider* is not supplied).
        model: Chat model ID (e.g. ``"gpt-4o-mini"``).
        system_prompt: System instructions for the agent.
        tools: Tool definitions from the agent (may include local handlers
            and/or webhook URLs).
        tool_executor: A ``ToolExecutor`` instance for running tools.
        llm_provider: An optional custom :class:`LLMProvider`.  When omitted
            an :class:`OpenAILLMProvider` is created using *openai_key* and
            *model* (backward compatible).
    """

    def __init__(
        self,
        openai_key: str,
        model: str,
        system_prompt: str,
        tools: list[dict] | None = None,
        tool_executor=None,
        llm_provider: LLMProvider | None = None,
        metrics=None,
        event_bus=None,
    ) -> None:
        if llm_provider is not None:
            self._provider = llm_provider
        else:
            self._provider = OpenAILLMProvider(api_key=openai_key, model=model)

        self._system_prompt = system_prompt
        self._tools = tools
        self._tool_executor = tool_executor
        self._metrics = metrics
        self._event_bus = event_bus
        self._model = model
        # Resolve the provider key for cost attribution. Prefer the
        # ``provider_key`` ClassVar declared by wrapper classes (stable,
        # matches ``pricing.py``); fall back to the legacy ``__name__``
        # strip for custom user-defined providers.
        if llm_provider is not None:
            cls = type(llm_provider)
            explicit = getattr(cls, "provider_key", None)
            if explicit:
                self._provider_name = explicit
            else:
                raw = cls.__name__.lower()
                for suffix in ("llmprovider", "provider", "llm"):
                    raw = raw.replace(suffix, "")
                self._provider_name = raw or "custom"
        else:
            self._provider_name = "openai"

        # Build OpenAI-format tool definitions (without handler/webhook_url)
        self._openai_tools: list[dict] | None = None
        if tools:
            self._openai_tools = []
            for t in tools:
                fn = {
                    "name": t["name"],
                    "description": t.get("description", ""),
                    "parameters": t.get("parameters", {"type": "object", "properties": {}}),
                }
                self._openai_tools.append({"type": "function", "function": fn})

        # Map tool name -> original tool dict (for handler/webhook_url lookup)
        self._tool_map: dict[str, dict] = {}
        if tools:
            for t in tools:
                self._tool_map[t["name"]] = t

    async def run(
        self,
        user_text: str,
        history: list[dict],
        call_context: dict,
        hook_executor=None,
        hook_ctx=None,
    ) -> AsyncGenerator[str, None]:
        """Stream LLM response tokens, handling tool calls automatically.

        Builds messages from history + user_text, streams via the configured
        :class:`LLMProvider`.  If the model emits tool calls, executes them
        via ``ToolExecutor`` and re-submits to the LLM until a text response
        is produced.

        Args:
            user_text: The user's latest transcribed utterance.
            history: Conversation history as ``[{role, text, timestamp}]``.
            call_context: Dict with ``call_id``, ``caller``, ``callee``.
            hook_executor: Optional :class:`PipelineHookExecutor` — when
                supplied, ``before_llm`` runs against the messages list
                before each provider call, and ``after_llm`` runs against
                the final assistant text once streaming completes.
            hook_ctx: Optional :class:`HookContext` — required when
                ``hook_executor`` is supplied.

        Yields:
            Text tokens as they arrive from the LLM.
        """
        messages = self._build_messages(history, user_text)
        # before_llm hook runs once on the initial message list. Subsequent
        # tool-call iterations re-submit augmented messages and skip the
        # hook (running the hook on every iteration would let a poorly
        # written hook trigger an infinite re-write loop).
        # Note: ``after_llm`` rewriting is only meaningful pre-TTS, so when
        # the hook is configured we buffer all tokens, run the hook, and
        # yield the (possibly rewritten) text as a single chunk. Without
        # the hook, streaming continues token-by-token as before.
        has_after_llm = (
            hook_executor is not None
            and hook_ctx is not None
            and getattr(hook_executor, "_hooks", None) is not None
            and getattr(hook_executor._hooks, "after_llm", None) is not None
        )
        if hook_executor is not None and hook_ctx is not None:
            messages = await hook_executor.run_before_llm(messages, hook_ctx)
        # Accumulate yielded text across iterations for after_llm hook.
        all_emitted_text: list[str] = []

        # Loop to handle tool calls — the LLM may call tools multiple times
        max_iterations = 10
        for iteration in range(max_iterations):
            tool_calls_accumulated: dict[int, dict] = {}
            text_parts: list[str] = []
            has_tool_calls = False

            # Open a span around the provider streaming call. Kept as an
            # explicit __enter__/__exit__ (rather than ``with``) because we
            # need to ``yield`` from inside the span which ``with`` + async
            # generators makes awkward.
            _span_cm = start_span(
                SPAN_LLM,
                {
                    "patter.llm.iteration": iteration,
                    "patter.llm.history_size": len(history),
                    "patter.call.id": call_context.get("call_id", ""),
                },
            )
            _span_cm.__enter__()
            try:
                async for chunk in self._provider.stream(messages, self._openai_tools):
                    chunk_type = chunk.get("type")

                    if chunk_type == "text":
                        content = chunk.get("content", "")
                        if content:
                            text_parts.append(content)
                            if self._event_bus is not None:
                                self._event_bus.emit(
                                    "llm_chunk",
                                    {"text": content, "iteration": iteration},
                                )
                            if has_after_llm:
                                # Buffer; yield after the after_llm hook runs.
                                all_emitted_text.append(content)
                            else:
                                yield content

                    elif chunk_type == "usage":
                        if self._metrics is not None:
                            self._metrics.record_llm_usage(
                                provider=self._provider_name,
                                model=self._model,
                                input_tokens=chunk.get("input_tokens", 0),
                                output_tokens=chunk.get("output_tokens", 0),
                                cache_read_tokens=chunk.get("cache_read_input_tokens", 0),
                                cache_write_tokens=chunk.get("cache_creation_input_tokens", 0),
                            )

                    elif chunk_type == "tool_call":
                        has_tool_calls = True
                        idx = chunk["index"]
                        if idx not in tool_calls_accumulated:
                            tool_calls_accumulated[idx] = {
                                "id": "",
                                "name": "",
                                "arguments": "",
                            }
                            # Emit tool_call_started the first time we see
                            # a given index. ``args`` may still be empty —
                            # streamed tool args arrive incrementally.
                            if self._event_bus is not None:
                                self._event_bus.emit(
                                    "tool_call_started",
                                    {
                                        "index": idx,
                                        "name": chunk.get("name") or "",
                                        "args": chunk.get("arguments") or "",
                                    },
                                )
                        if chunk.get("id"):
                            tool_calls_accumulated[idx]["id"] = chunk["id"]
                        if chunk.get("name"):
                            tool_calls_accumulated[idx]["name"] = chunk["name"]
                        if chunk.get("arguments"):
                            tool_calls_accumulated[idx]["arguments"] += chunk["arguments"]
            finally:
                _span_cm.__exit__(None, None, None)

            # If no tool calls, we're done
            if not has_tool_calls:
                if has_after_llm:
                    final_text = "".join(all_emitted_text)
                    rewritten = await hook_executor.run_after_llm(final_text, hook_ctx)
                    if rewritten:
                        yield rewritten
                return

            # Execute tool calls and add results to messages
            assistant_msg: dict = {"role": "assistant", "content": "".join(text_parts) or None}
            assistant_tool_calls = []
            for idx in sorted(tool_calls_accumulated.keys()):
                tc = tool_calls_accumulated[idx]
                assistant_tool_calls.append({
                    "id": tc["id"],
                    "type": "function",
                    "function": {
                        "name": tc["name"],
                        "arguments": tc["arguments"],
                    },
                })
            assistant_msg["tool_calls"] = assistant_tool_calls
            messages.append(assistant_msg)

            for tc_data in assistant_tool_calls:
                tool_name = tc_data["function"]["name"]
                try:
                    arguments = json.loads(tc_data["function"]["arguments"])
                except json.JSONDecodeError:
                    arguments = {}

                result = await self._execute_tool(tool_name, arguments, call_context)
                messages.append({
                    "role": "tool",
                    "tool_call_id": tc_data["id"],
                    "content": result,
                })

            # Re-submit to LLM with tool results — next iteration will
            # either produce text or more tool calls

        logger.warning("LLM loop hit max iterations (%d)", max_iterations)

    async def _execute_tool(
        self, tool_name: str, arguments: dict, call_context: dict
    ) -> str:
        """Execute a tool via ToolExecutor."""
        tool_def = self._tool_map.get(tool_name, {})
        handler = tool_def.get("handler")
        webhook_url = tool_def.get("webhook_url", "")

        if self._tool_executor is not None:
            return await self._tool_executor.execute(
                tool_name=tool_name,
                arguments=arguments,
                call_context=call_context,
                webhook_url=webhook_url,
                handler=handler,
            )

        return json.dumps({"error": f"No executor available for tool '{tool_name}'"})

    def _build_messages(
        self, history: list[dict], user_text: str
    ) -> list[dict]:
        """Build OpenAI messages array from conversation history."""
        messages: list[dict] = [
            {"role": "system", "content": self._system_prompt},
        ]
        for entry in history:
            role = entry.get("role", "user")
            text = entry.get("text", "")
            if role == "assistant":
                messages.append({"role": "assistant", "content": text})
            else:
                messages.append({"role": "user", "content": text})

        # Add the current user message
        messages.append({"role": "user", "content": user_text})
        return messages
